### ISSUE TITLE
improve cunit code coverage for lib/ directory

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#2037]
The functions and such in the lib/ directory are mostly pretty generic, they don't depend on a live cyrus instance to function.  So we should be able to get pretty good test coverage of these just using cunit tests.  We should aim for ~90% coverage for this stuff, because it's easy to achieve and verifies the deepest core of cyrus.

**Comment by elliefm** [comment]
Coverage summary for lib/ directory at time of writing: [coverage-lib.pdf](https://github.com/cyrusimap/cyrus-imapd/files/1093657/coverage-lib.pdf)

**Comment by rsto** [comment]
That's a great idea, @elliefm ! As a minimum, I can work on improving charset coverage. I also vouch for kicking out cyrusdb_lmdb from master and 3.0, it's broken. We still can resurrect it from the repo, if necessary (what do you think, @brong ?)

**Comment by brong** [comment]
Agree with drop lmdb


On Thu, 22 Jun 2017, at 23:30, Robert Stepanek wrote:
> That's a great idea, @elliefm[1] ! As a minimum, I can work on
> improving charset coverage. I also vouch for kicking out cyrusdb_lmdb
> from master and 3.0, it's broken. We still can resurrect it from the
> repo, if necessary (what do you think, @brong[2] ?)> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[3], or mute the thread[4].> 

--
  Bron Gondwana
  brong@fastmail.fm



Links:

  1. https://github.com/elliefm
  2. https://github.com/brong
  3. https://github.com/cyrusimap/cyrus-imapd/issues/2037#issuecomment-310380600

**Comment by elliefm** [comment]
Is the lmdb brokenness something that could be reproduced from cunit tests?  Currently the `aaa-db.testc` suite passes all tests for backend=lmdb.  If it's reasonably doable, it'd be good to have some failing tests for it, so we don't trip over the same issues if we bring it back someday.

**Comment by brong** [comment]
I'm really not sure :(  It might require multiple threads writing at
once.
Bron.


On Fri, 23 Jun 2017, at 11:06, elliefm wrote:
> Is the lmdb brokenness something that could be reproduced from cunit
> tests?  Currently the aaa-db.testc suite passes all tests for
> backend=lmdb.  If it's reasonably doable, it'd be good to have some
> failing tests for it, so we don't trip over the same issues if we
> bring it back someday.> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].> 

--
  Bron Gondwana
  brong@fastmail.fm



Links:

  1. https://github.com/cyrusimap/cyrus-imapd/issues/2037#issuecomment-310542522

**Comment by rsto** [comment]
Yes, we couldn't reproduce the issue into a well-defined chain of commands (other than: change the code, pass all our tests, deploy it on future, see how brongs dbs corrupt over time). It's almost certainly in the way we treat writes/deletes in nested transactions.

But: there also was the issue that lmdb dbs tend to grow and grow, which brought the operational issue of recompacting without a stop-the-world lock. This wouldn't change, even if we get to fix the corrupt db issue.

**Comment by rjbs** [comment]
:robot: I'm closing this issue by automation.

This issue has been sitting untouched for over seven years.  We're trying to get a good handle on our open issues and pull requests, so that we can reliably use them as a backlog of things we hope to work on.  If you are confident that this issue should be reconsidered, reopen it and explain why.

Remember, though, that there are only so many developers working on Cyrus, and we'd like to focus on the most important and/or low-hanging work.

Before you reopen this issue, please check that this issue is still present in the latest release of Cyrus, or the main development branch. An issue is considerably more likely to get resolved if you can provide a pull request which reproduces the issue in a Cassandane test.
